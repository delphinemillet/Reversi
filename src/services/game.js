import { WHITE, BLACK, EMPTY, UNKNOWN } from "../constants/gameStates";

export default {
  cols: Array(...Array(8)),
  rows: Array(...Array(8)),
  state: [],

  // INITIALIZATION
  initGame() {
    this.rows.map((row, i) => {
      this.state[i] = []
      this.rows.map(() => this.state[i].push(EMPTY))
    })

    this.setState(3, 3, WHITE)
    this.setState(3, 4, BLACK)
    this.setState(4, 3, BLACK)
    this.setState(4, 4, WHITE)

    this.setPlayer(BLACK)
  },

  // TILE
  isInsideBounds(value) {
    return value >=0 && value <= 7
  },

  setState(i, j, state) {
    this.state[i].splice(j, 1, state)
  },

  getState(i, j) {
    return this.isInsideBounds(i) && this.isInsideBounds(j) ? this.state[i][j] : UNKNOWN
  },

  isValid(i, j) {
    const opponent = this.getOpponent()
    return this.getState(i, j) === EMPTY
      ? this.getState(i - 1, j) === opponent
        || this.getState(i + 1, j) === opponent
        || this.getState(i + 1, j) === opponent
        || this.getState(i, j - 1) === opponent
        || this.getState(i, j + 1) === opponent
      : false
  },

  // PLAYERS
  setPlayer(player) {
    this.currentPlayer = player
  },

  getOpponent() {
    return this.currentPlayer === BLACK ? WHITE : BLACK
  },

  // GAME
  play(i, j) {
    if(this.isValid(i, j)) {
      this.setState(i, j, this.currentPlayer)
      this.setPlayer(this.getOpponent())
    }
  }
};