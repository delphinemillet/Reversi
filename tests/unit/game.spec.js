
import game from "../../src/services/game"

describe('game service', () => {
  beforeEach(() => {
    game.initGame()
  })

  describe('initGame()', () => {
    it("should init the board", () => {
      expect(game.state).toMatchSnapshot()
    })
    it("should set the current player to BLACK", () => {
      expect(game.currentPlayer).toEqual('BLACK')
    })
  })

  describe('setState()', () => {
    it("should update the tile state on a given position", () => {
      game.setState(0, 0, 'value')
      expect(game.state[0][0]).toEqual('value')
    })
  })

  describe('getState()', () => {
    it("should return the tile state for a given position", () => {
      game.getState(0, 0)
      expect(game.state[0][0]).toEqual('EMPTY')
      expect(game.state[3][3]).toEqual('WHITE')
    })
  })

  describe('isValid()', () => {
    it('should return false if tile is not empty', () => {
      expect(game.isValid(3,3)).toBeFalsy()
    })
    it('should return true if tile is empty and adjacent tile is occupied by opponent', () => {
      expect(game.isValid(3,2)).toBeTruthy()
      expect(game.isValid(2,3)).toBeTruthy()
      expect(game.isValid(5,4)).toBeTruthy()
      expect(game.isValid(4,5)).toBeTruthy()
    })
    it('should return false if tile is empty but not adjacent to opponent tile', () => {
      expect(game.isValid(3,5)).toBeFalsy()
      expect(game.isValid(2,2)).toBeFalsy()
      expect(game.isValid(0,0)).toBeFalsy()
      expect(game.isValid(7,7)).toBeFalsy()
    })
  })

  describe('setPlayer()', () => {
    it("should set the current player", () => {
      game.setPlayer("toto")
      expect(game.currentPlayer).toEqual("toto")
    })
  })

  describe('getOpponent()', () => {
    it("should return the BLACK'S opponent", () => {
      expect(game.getOpponent()).toEqual("WHITE")
    })
    it("should return the WHITE'S opponent", () => {
      game.setPlayer('WHITE')
      expect(game.getOpponent()).toEqual("BLACK")
    })
  })

  describe('play()', () => {
    it('should do nothing if move is not valid', () => {
      game.play(0,0)
      expect(game.state).toMatchSnapshot()
      expect(game.currentPlayer).toEqual('BLACK')
    })
    it('should set the player\'s tile and update the current player', () => {
      game.play(3,2)
      expect(game.state).toMatchSnapshot()
      expect(game.currentPlayer).toEqual('WHITE')
    })
  })
})
