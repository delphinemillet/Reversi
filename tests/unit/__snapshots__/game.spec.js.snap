// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game service initGame() should init the board 1`] = `
Array [
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "WHITE",
    "BLACK",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "BLACK",
    "WHITE",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
]
`;

exports[`game service play() should do nothing if move is not valid 1`] = `
Array [
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "WHITE",
    "BLACK",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "BLACK",
    "WHITE",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
]
`;

exports[`game service play() should set the player's tile and update the current player 1`] = `
Array [
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "BLACK",
    "WHITE",
    "BLACK",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "BLACK",
    "WHITE",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
  Array [
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
    "EMPTY",
  ],
]
`;
